src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -205,26 +203,264 @@ bool tryBindAsHandlerType([NotNullWhen(true)] out BoundInterpolatedString? resul
+                }
+            }
+
+            Debug.Assert(partsArrayBuilder.Count == stack.Count + 1);
+            Debug.Assert(partsArrayBuilder.Count >= 2);
+
+            if (isConstant ||
+                (partsCount <= 4 && partsArrayBuilder.All(static parts => AllInterpolatedStringPartsAreStrings(parts))))
+            {
+                // This is case 1 and 2. Let the standard machinery handle it
+                stack.Free();
+                partsArrayBuilder.Free();
+                return false;
+            }
+
+            // Parts were added to the array from right to left, but lexical order is left to right.
+            partsArrayBuilder.ReverseContents();
+
+            // Case 3. Bind as handler.
+            var (appendCalls, data) = BindUnconvertedInterpolatedPartsToHandlerType(
+                binaryOperator.Syntax,
+                partsArrayBuilder.ToImmutableAndFree(),
+                interpolatedStringHandlerType,
+                diagnostics,
+                isHandlerConversion: false,
+                additionalConstructorArguments: default,
+                additionalConstructorRefKinds: default);
+
+            // Now that the parts have been bound, reconstruct the binary operators.
+            convertedBinaryOperator = UpdateBinaryOperatorWithInterpolatedContents(stack, appendCalls, data, binaryOperator.Syntax, diagnostics);
+            stack.Free();
+            return true;
+        }
+
+        private BoundBinaryOperator UpdateBinaryOperatorWithInterpolatedContents(ArrayBuilder<BoundBinaryOperator> stack, ImmutableArray<ImmutableArray<BoundExpression>> appendCalls, InterpolatedStringHandlerData data, SyntaxNode rootSyntax, BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(appendCalls.Length == stack.Count + 1);
+            var @string = GetSpecialType(SpecialType.System_String, diagnostics, rootSyntax);
+            var signature = new BinaryOperatorSignature(BinaryOperatorKind.StringConcatenation, @string, @string, @string);
+
+            var bottomOperator = stack.Pop();
+            var result = createBinaryOperator(bottomOperator, createInterpolation(bottomOperator.Left, appendCalls[0]), rightIndex: 1);
+
+            for (int i = 2; i < appendCalls.Length; i++)
+            {
+                result = createBinaryOperator(stack.Pop(), result, rightIndex: i);
+            }
+
+            return result.Update(BoundBinaryOperator.UncommonData.WithInterpolatedStringHandlerData(data));
+
+            BoundBinaryOperator createBinaryOperator(BoundBinaryOperator original, BoundExpression left, int rightIndex)
+                => new BoundBinaryOperator(
+                    original.Syntax,
+                    BinaryOperatorKind.StringConcatenation,
+                    left,
+                    createInterpolation(original.Right, appendCalls[rightIndex]),
+                    original.ConstantValue,
+                    signature.Method,
```

Comment:
Why are we building a `BinaryOperatorSignature` instance, only to use the `Method` (which is `null`) and `ConstrainedToTypeOpt` (also `null`?)?